core/celo_multi_gaspool.go

```diff
@@ -0,0 +1,57 @@
+package core
+
+import (
+	// "fmt"
+	// "math"
+
+	"github.com/celo-org/celo-blockchain/common"
+)
+
+type FeeCurrency = common.Address
+
+// MultiGasPool tracks the amount of gas available during execution
+// of the transactions in a block per fee currency. The zero value is a pool
+// with zero gas available.
+type MultiGasPool struct {
+	pools       map[FeeCurrency]*GasPool
+	defaultPool *GasPool
+}
+
+type FeeCurrencyLimitMapping = map[FeeCurrency]float64
+
+func NewMultiGasPool(
+	block_gas_limit uint64,
+	whitelist []FeeCurrency,
+	defaultLimit float64,
+	limitsMapping FeeCurrencyLimitMapping,
+) MultiGasPool {
+	pools := make(map[FeeCurrency]*GasPool, len(whitelist))
+
+	for i := range whitelist {
+		currency := whitelist[i]
+		fraction, ok := limitsMapping[currency]
+		if !ok {
+			fraction = defaultLimit
+		}
+
+		pools[currency] = new(GasPool).AddGas(
+			uint64(float64(block_gas_limit) * fraction),
+		)
+	}
+
+	// A special case for CELO which doesn't have a limit
+	celoPool := new(GasPool).AddGas(block_gas_limit)
+
+	return MultiGasPool{
+		pools:       pools,
+		defaultPool: celoPool,
+	}
+}
+
+func (mgp MultiGasPool) PoolFor(feeCurrency *FeeCurrency) *GasPool {
+	if feeCurrency == nil || mgp.pools[*feeCurrency] == nil {
```

Comment:
I don't think we need the `defaultPool` here, as this is checked in `blockState.gasPool` and there should never be an unregistered fee currency here.

But I guess this is mostly to make the code return a `*GasPool`, which is fine.

Reply:
Yes, that's mostly to avoid errors or additional checks in case of `nil` as `feeCurrency`, as I didn't want that code to be dependent on where it's used and rely on checks made elsewhere.

Reply:
Makes sense!